### PR TITLE
Allow fixed number of steps on HMC

### LINKED ISSUE #1356: Allow a fixed num_steps in HMC
It would be nice to allow a fixed num_steps in HMC. If the trajectory length is fixed and step size is too small, num_steps will turn out to be quite large. This will be helpful for #1355 and #1353.

### PR DESCRIPTION
**Pull request by themrzmaster**
Attempts to close #1356 .

My first PR here, I am not sure if I did it the right way

@fehiepsi tagging you because of the mentioned issue :)

### CODE REVIEW

File: test/infer/test_mcmc.py [hunk 1]
```diff
@@ -1106,3 +1106,27 @@ def eight_schools(J, sigma):
     rng_key = random.PRNGKey(0)
     mcmc.run(rng_key, J, sigma, extra_fields=("potential_energy",))
     assert mcmc.get_samples()["scores"].shape == (200, 0, 8)
+
+
+def test_fixed_num_steps():
+    data = dict()
+    data['x'] = np.array([0.0970,2.1020,0.5840,1.0394,3.4375,1.3102,1.2863,4.5382,1.2539,2.9319,4.7777,4.5937,4.0403,0.7749,1.8342,1.5008,3.9557,1.6095,0.5602,4.7997,1.6436,4.5236,1.9404,1.1508,3.1447,3.3551,0.2669,3.2483,2.5293,0.1596,2.8992,2.5016,4.7589,2.5648,1.9871,0.3201,4.4863,0.5675,0.8897,0.5689,0.5621,0.5701,0.7998,0.3166,4.8490,2.7849,2.6858,4.2393,1.4939,1.7230,0.4202,0.3591,2.5912,3.9753,3.5417,0.3198,2.7192,1.9892,1.9525,4.7293,1.7719,3.7889,2.3613,4.6994,2.0155,3.8840,0.8289,2.1164,1.3342,1.4850,0.5512,0.3919,2.4001,3.1949,4.9759,1.4158,0.6088,2.1599,0.3643,0.6800,2.4031,1.7706,3.7669,4.1080,4.9707,1.9139,4.7615,3.3075,3.6872,4.0946,0.3447,2.7292,0.2768,3.7641,3.3987,0.1833,4.9353,1.9730,2.2034,2.8598])
```
Comment (themrzmaster) [inline test/infer/test_mcmc.py:1113]: I used the example over here https://github.com/pyro-ppl/numpyro/issues/1355
Do we have a better way to test?

**Review by fehiepsi — COMMENTED** [review]
Thanks @themrzmaster!  I like the pattern. Just a minor comment to cleanup the hmc state.

File: numpyro/infer/hmc.py [hunk 2]
```diff
@@ -608,9 +602,9 @@ def __init__(
         potential_fn=None,
         kinetic_fn=None,
         step_size=1.0,
+        num_steps=None,
```
Comment (fehiepsi) [inline numpyro/infer/hmc.py:606]: I think we need to move this to near `trajectory_length` argument and force keyword arguments for it to make it less likely to break the current users' code: `... target_accept_prob=0.8, *, num_steps=None, trajectory_length=2 * pi`. Could you also add docstring for it like in `hmc`?

**Review by fehiepsi — APPROVED** [review]

**Comment by fehiepsi** [comment]
@themrzmaster look like you need to run `make format`

**Comment by fehiepsi** [comment]
Sorry for troublesome, you might need to use latest black version.

**Comment by themrzmaster** [comment]
I guess the tests failures are not related.

**Comment by fehiepsi** [comment]
That's right. I made the fixes for them in https://github.com/pyro-ppl/numpyro/pull/1371.

**Comment by fehiepsi** [comment]
Hi @themrzmaster, could you merge master branch?

**Comment by themrzmaster** [comment]
done @fehiepsi !

**Review by fehiepsi — APPROVED** [review]
Thanks!

**Pull request merged**